src/Emogrifier/CssConcatenator.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+namespace Pelago\Emogrifier;
+
+/**
+ * Facilitates building a CSS string by appending rule blocks one at a time, checking whether the media query,
+ * selectors, or declarations block are the same as those from the preceding block and combining blocks in such cases.
+ *
+ * Example:
+ *  $concatenator = new CssConcatenator();
+ *  $concatenator->append(['body'], 'color: blue;');
+ *  $concatenator->append(['body'], 'font-size: 16px;');
+ *  $concatenator->append(['p'], 'margin: 1em 0;');
+ *  $concatenator->append(['ul', 'ol'], 'margin: 1em 0;');
+ *  $concatenator->append(['body'], 'font-size: 14px;', '@media screen and (max-width: 400px)');
+ *  $concatenator->append(['ul', 'ol'], 'margin: 0.75em 0;', '@media screen and (max-width: 400px)');
+ *  $css = $concatenator->getCss();
+ *
+ * `$css` (if unminified) would contain the following CSS:
+ * ` body {
+ * `   color: blue;
+ * `   font-size: 16px;
+ * ` }
+ * ` p, ul, ol {
+ * `   margin: 1em 0;
+ * ` }
+ * ` @media screen and (max-width: 400px) {
+ * `   body {
+ * `     font-size: 14px;
+ * `   }
+ * `   ul, ol {
+ * `     margin: 0.75em 0;
+ * `   }
+ * ` }
+ *
+ * @author Jake Hotson <jake.github@qzdesign.co.uk>
+ */
+class CssConcatenator
+{
+    /**
+     * Array of media rules in order.  Each element is an object with the following properties:
+     * - string `media` - The media query string, e.g. "@media screen and (max-width:639px)", or an empty string for
+     *   rules not within a media query block;
+     * - stdClass[] `ruleBlocks` - Array of rule blocks in order, where each element is an object with the following
+     *   properties:
+     *   - int[] `selectorsAsKeys` - Array whose keys are selectors for the rule block (values are of no significance);
+     *   - string `declarationsBlock` - The property declarations, e.g. "margin-top: 0.5em; padding: 0".
+     *
+     * @var stdClass[]
+     */
+    private $mediaRules = [];
```

Comment:
I've noticed another way we could simplify this: We could use the media as array key for `$mediaRules`, effectively removing another dimension from the array, and allowing us to simplify the lookups. @JakeQZ What do you think?

Reply:
This wouldn't work if there were multiple media rules for the same media query, interspersed with other rules, e.g.
```
blockquote::before {
  font-size: 16px;
}
@media (max-width: 479px) {
  blockquote::before {
    font-size: 14px;
  }
}
.smallprint blockquote::before {
  font-size: 12px;
}
@media (max-width: 479px) {
  .smallprint blockquote::before {
    font-size: 10px;
  }
}
```
The final `10px` declaration would end up in the first media rule and be overridden by the `12px` declaration when it shouldn't.

Reply:
Ah, makes sense. Is this case already covered by the tests?

Reply:
No, currently they just test combining or not combining pairs of rules.  I'll add a test for not combining rules (which otherwise would be) when there is some other rule between.